prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst

```diff
@@ -0,0 +1,627 @@
+                         Calibrate/Train
+                                |
+    —--------------------------------------------------------
+    |                      convert_pt2e                     |
+    —--------------------------------------------------------
+                                |
+                    Reference Quantized Model
+                                |
+    —--------------------------------------------------------
+    |                        Lowering                       |
+    —--------------------------------------------------------
+                                |
+            Executorch, or Inductor, or <Other Backends>
+
+An existing quantizer object defined for QNNPack/XNNPack is located in
+`QNNPackQuantizer <https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py>`__.
+Taking QNNPackQuantizer as an example, the overall Quantization 2.0 flow could be:
+
+::
+
+    import torch
+    import torch._dynamo as torchdynamo
+    from torch.ao.quantization._quantize_pt2e import convert_pt2e, prepare_pt2e
+    import torch.ao.quantization._pt2e.quantizer.qnnpack_quantizer as qq
+
+    class M(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.linear = torch.nn.Linear(5, 10)
+
+        def forward(self, x):
+            return self.linear(x)
+
+    example_inputs = (torch.randn(1, 5),)
+    model = M().eval()
+
+    # Step 1: Trace the model into an FX graph of flattened ATen operators
+    exported_graph_module, guards = torchdynamo.export(
+        model,
+        *copy.deepcopy(example_inputs),
+        aten_graph=True,
+    )
+
+    # Step 2: Insert observers or fake quantize modules
+    quantizer = qq.QNNPackQuantizer()
+    operator_config = qq.get_symmetric_quantization_config(is_per_channel=True)
+    quantizer.set_global(operator_config)
+    prepared_graph_module = prepare_pt2e_quantizer(exported_graph_module, quantizer)
+
+    # Step 3: Quantize the model
+    convered_graph_module = convert_pt2e(prepared_graph_module)
+
+    # Step 4: Lower Reference Quantized Model into the backend
+
+Inside the Quantizer, we will use the ``QuantizationAnnotation API``
+to convey user's intent for what quantization spec to use and how to
+observe certain tensor values in the prepare step. Now, we will have a step-by-step
+tutorial for how to use the ``QuantizationAnnotation API`` with different types of
+``QuantizationSpec``.
+
+1. Annotate common operator patterns
+--------------------------------------------------------
+
+`QuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L38>`__
+is used to annotate common operators like ``conv2d``. It allows user to specify how to quantize
+input tensors and output tensor of this operator which includes parameters of ``observer type``, ``dtype``,
+``quant_min``, and ``quant_max`` etc.
+
+::
```

Comment:
I feel we do not necessarily need to show this code, but instead it will be helpful to start with a use case here, let's say, we can talk about the use case of a quantized conv2d, it should go through these feel steps:
1. identify conv2d pattern
2. derive the quantization spec for input and output Tensors
3. annotate the nodes in the pattern properly

you can also talk about how to annotate conv - relu pattern in this section

Reply:
maybe start with add / add - relu as an example here, since conv requires DerivedQuantizationSpec for bias

Reply:
you can follow the steps in https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit#heading=h.it9h4gjr7m9g for each section that introduces a new quantization spec I think

Reply:
> I feel we do not necessarily need to show this code, but instead it will be helpful to start with a use case here, let's say, we can talk about the use case of a quantized conv2d, it should go through these feel steps:

Hi @jerryzh168, Thanks for the suggestions. I have updated the document by elaboration more about use case and steps to use each ``QuantizationSpec``.  Please help to take a look again.
One question is about ``DerivedQuantizationSpec``, I think this [document ](https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit#) didn't introduce motivation clearly. Since we generally use `PlaceholderObserver` with `QuantizationSpec` for bias node, why we have to use ``DerivedQuantizationSpec``?

Reply:
I see, this is for the Tensors whose quantization parameters are derived from other Tensors, for example bias, the `PlaceholderObserver` that we used previously did not express this relation correctly

Reply:
Hi @jerryzh168. Thanks for the comment. Will the [`bias_quantization_spec`](https://github.com/pytorch/pytorch/blob/a02a58d862b00a7aa410149e8cf276b8d0ca3462/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L183-L185) in `QNNPackQuantizer` updated to use `DerivedQuantizationSpec`? If so, does it mean there will be the `fp32_bias -> quant_to_int32 -> dequant -> conv` pattern instead of `fp32_bias -> conv` after `convert_pt2e`?

Reply:
yeah I think it will be updated eventually, but I don't know when we plan to do that yet, @kimishpatel would have more context on the timeline